src/file_pipe_log/file_mgr.rs

```diff
@@ -0,0 +1,674 @@
+            (purged_files.first_seq, purged_files.len());
+        let logical_purged_path_list: Vec<PathId> =
+            purged_files.fds.iter().map(|f| f.path_id).collect();
+        // If there exists several stale files for purging in `self.stale_files`,
+        // we should check and remove them to avoid the `size` of whole files
+        // beyond `self.capacity`.
+        let clear_list = {
+            self.stale_files.append(purged_files);
+            self.stale_files
+                .purge(self.capacity.saturating_sub(self.active_files.len()))
+        };
+        for (i, fd) in clear_list.fds.iter().enumerate() {
+            #[cfg(feature = "failpoints")]
+            {
+                let remove_skipped = || {
+                    fail::fail_point!("file_pipe_log::remove_file_skipped", |_| true);
+                    false
+                };
+                if remove_skipped() {
+                    continue;
+                }
+            }
+            let seq = clear_list.first_seq + i as FileSeq;
+            let file_id = FileId {
+                queue: self.queue,
+                seq,
+            };
+            // If seq less than the first seq of `purged_files`, it should be a renamed
+            // stale file, named with `.raftlog.stale` suffix.
+            let path = if seq < logical_purged_first_seq {
+                file_id.build_stale_file_path(&self.paths[fd.path_id])
+            } else {
+                file_id.build_file_path(&self.paths[fd.path_id])
+            };
+            self.file_system.delete(&path)?;
+        }
+        // Meanwhile, the new supplemented files from `self.active_files` should
+        // be RENAME to stale files with `.raftlog.stale` suffix, to reduce the
+        // unnecessary recovery timecost when RESTART.
+        {
+            let mv_first_seq = std::cmp::max(
+                logical_purged_first_seq,
+                clear_list.span().0 + clear_list.len() as FileSeq,
+            );
+            let mv_total_len = std::cmp::min(logical_purged_count, self.stale_files.len());
+            for seq in mv_first_seq..mv_first_seq + mv_total_len as FileSeq {
+                let file_id = FileId {
+                    queue: self.queue,
+                    seq,
+                };
+                let path_id = logical_purged_path_list[(seq - logical_purged_first_seq) as usize];
+                let path = file_id.build_file_path(&self.paths[path_id]);
+                let stale_path = file_id.build_stale_file_path(&self.paths[path_id]);
+                self.file_system.reuse(&path, &stale_path)?;
+            }
+        }
+        Ok(logical_purged_count)
+    }
+
+    /// Initialize current file collection by preparing several
+    /// stale files for later log recycling in advance.
+    ///
+    /// Attention, this function only makes sense when
+    /// `Config.enable-log-recycle` is true.
+    pub fn initialize(&mut self, target_file_size: usize) -> Result<()> {
```

Comment:
I don't know why this function is so long a gain. Why doesn't this work?
```rust
for _ in self.active_files.len() + self.stale_files.len() .. self.capacity {
  let f = build_stale_files(self.stale_files.span().1 + 1);
  self.stale_files.push_back(f);
}
```

Reply:
For this, the whole design here follows the prerequisite --
+ `seqno` is consecutive in `stale_files` and `active_files`;
+ `stale_files.last_seq` + 1 == `active_files.first`.

So, if we start the engine on following special scenario:
```Rust
// Existed stale files: [16, ..., 21]
// Existed active files: [45, ..., 80]
```
we should calibrate the `seqno` in the existed stale files, to make all `seqno` consecutive:
```Rust
// After initializing =>
// stale files: [35, ..., 40][41, ..., 44] (this part is newly added by Initialize().)
// active files: [45, ..., 80]
```

Reply:
> stale_files.last_seq + 1 == active_files.first

No need.